### PR TITLE
Poseidon hash

### PR DESCRIPTION
**Pull request by igamigo**
- Adds Poseidon implementation
- Adds parameter files for BLS12381's field 
- Tests based on Starkware's implementation values in order to confirm both implementations yield similar results

### CODE REVIEW

File: math/src/unsigned_integer/element.rs [hunk 1]
```diff
@@ -8,7 +8,6 @@ use crate::unsigned_integer::traits::IsUnsignedInteger;
 use std::fmt::Debug;
 
 pub type U384 = UnsignedInteger<6>;
-pub type U256 = UnsignedInteger<4>;
```
Comment (jrchatruc) [inline math/src/unsigned_integer/element.rs:11]: Should this line be removed?

Comment (igamigo) [reply math/src/unsigned_integer/element.rs:11]: Yeah, I had added it myself but then it turned out I did not need it so I removed it. However it turns out someone else had pushed the same line to main yesterday and I had missed that after merging. Fixed, thanks.

**Review by jrchatruc — APPROVED** [review]

**Review by ilitteri — APPROVED** [review]
LGTM

**Review by MauroToscano — APPROVED** [review]

File: crypto/src/hash/poseidon/mod.rs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+            TestFieldElement::new(U384::from(
+                "16861759ea5568dd39dd92f9562a30b9e58e2ad98109ae4780b7fd8eac77fe8a",
+            )),
+            TestFieldElement::new(U384::from(
+                "13827681995D5ADFFFC8397A3D00425A3DA43F76ABF28A64E4AB1A22F275092B",
+            )),
+            TestFieldElement::new(U384::from(
+                "BA3956D2FAD4469E7F760A2277DC7CB2CAC75DC279B2D687A0DBE17704A8310",
+            )),
+        ];
```
[the remark this answers is not in the document]
Comment (igamigo) [reply crypto/src/hash/poseidon/mod.rs:220]: These were taken from Starkware's implementation in order to test against their results. Mentioned this and referenced the code on comments above the test mod.

File: crypto/src/hash/poseidon/mod.rs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        let expected = [
+            TestFieldElement::new(U384::from(
+                "1d30b34b465f8cddc8dc468f137891659c7e32b510cf41cec3aac0b26741681d",
+            )),
+            TestFieldElement::new(U384::from(
+                "c445fa4dd2af583994272bede589b06b98fe9cd6d868bf718f6748ba6165620",
+            )),
+            TestFieldElement::new(U384::from(
+                "1ed95ae0ea03bb892691f5200fb5902957ac17b3466afa62be808682801f97f9",
+            )),
```
[the remark this answers is not in the document]
Comment (igamigo) [reply crypto/src/hash/poseidon/mod.rs:251]: These were taken from Starkware's implementation in order to test against their results. Mentioned this and referenced the code on comments above the test mod.

File: crypto/src/hash/poseidon/mod.rs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+            TestFieldElement::new(U384::from(
+                "18700783647721BB9AD092B176BBEB5348401C21132CCF83C30134DFAB5A2DEB",
+            )),
+            TestFieldElement::new(U384::from(
+                "1CC8856652601B3C81139AD5EC13E4A3A8F4A5DB242555521A09E002E7A10B2B",
+            )),
+            TestFieldElement::new(U384::from(
+                "3DCB1CEC811FC2D7401CA7B9B084D167F33B6983D4428C8E0534C9C3CECF46D",
+            )),
+        ];
```
[the remark this answers is not in the document]
Comment (igamigo) [reply crypto/src/hash/poseidon/mod.rs:289]: These were taken from Starkware's implementation in order to test against their results. Mentioned this and referenced the code on comments above the test mod.

**Pull request merged**